{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "18707189",
   "metadata": {},
   "source": [
    "# Workshop 5 - Automate BLAST Runs\n",
    "\n",
    "### Topics\n",
    "\n",
    "* About BLAST\n",
    "* Types of BLAST \n",
    "* BLAST Sequence Databases \n",
    "* Biopython's BLAST Tools \n",
    "    * Running BLAST over the Internet \n",
    "    * Saving BLAST output \n",
    "    * Parsing BLAST output \n",
    "    * Retrieving information from BLAST Record object \n",
    "* More Tools in Biopython \n",
    "    * Connecting with biological databases \n",
    "    * Going 3D\n",
    "    * Population Genetics\n",
    "    * Phylogenetics "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a8809ca",
   "metadata": {},
   "source": [
    "## About BLAST \n",
    "\n",
    "BLAST (basic local alignment search tool) is an algorithm and program for comparing primary biological sequence information, such as the amino-acid sequences of proteins or the nucleotides of DNA and/or RNA sequences. A BLAST search enables a researcher to compare a subject protein or nucleotide sequence (called a query) with a library or database of sequences, and identify database sequences that resemble the query sequence above a certain threshold. For example, following the discovery of a previously unknown gene in the mouse, a scientist will typically perform a BLAST search of the human genome to see if humans carry a similar gene; BLAST will identify sequences in the human genome that resemble the mouse gene based on similarity of sequence. \n",
    "\n",
    "## Types of BLAST\n",
    "\n",
    "* __Nucleotide-Nucleotide BLAST (blastn)__: given a DNA query, returns the most similar DNA sequences from the DNA database that the user specifies. \n",
    "* __Protein-Protein BLAST (blastp)__: given a protein query, returns the most similar protein sequences from the protein database that the user specifies. \n",
    "* __Translated Nucleotide-Protein BLAST (blastx)__: for identifying potential protein products encoded by a nucleotide query. \n",
    "* __Protein-Translated Nucleotide (tblastn)__: for identifying database sequences encoding proteins similar to the query. \n",
    "\n",
    "## BLAST Sequence Databases \n",
    "\n",
    "Nucleotide Database:\n",
    "\n",
    "* __nt (default)__: All GenBank + EMBL + DDBJ + PDB sequences, excluding sequences from PAT, EST, STS, GSS, WGS, TSA and phase 0, 1 or 2 HTGS sequences. Non-redundant, records with identical sequences collapsed into a single entry. \n",
    "* __rRNA/ITS databases__: A collection of four databases: a 16S Microbial rRNA sequences from NCBI's Targeted Loci Projects, an 18S and a 26S RNA rRNA databases for fungi, plus an ITS database for fungi. \n",
    "* __refseq_rna__: Curated (NM_, NR_) plus predicted (XM_, XR_) sequences from NCBI Reference Sequence Project. \n",
    "* ... and more\n",
    "\n",
    "Protein Databases:\n",
    "\n",
    "* __nr (default)__: Non-redundant GenBank CDS translations + RefSeq + PDB + SwissProt + PIR + PRF, excluding those in PAT, TSA, and env_nr.\n",
    "* __refseq_protein__: Protein sequences from NCBI Reference Sequence project. \n",
    "* __Landmark__: The landmark database includes proteomes from representative genomes spanning a wide taxonomic range.\n",
    "* ... and more\n",
    "\n",
    "For more information about BLAST programs and databases, please see NCBI's [How To BLAST Guide](https://ftp.ncbi.nlm.nih.gov/pub/factsheets/HowTo_BLASTGuide.pdf).\n",
    "\n",
    "## Biopython's BLAST Tools \n",
    "\n",
    "Most people's first introduction to running BLAST was probably via NCBI web-service. It can be difficult to deal with the volume of data generated by large runs, and to automate BLAST runs in general. Biopython provides several tools to make things easier. We will introduce a few here. \n",
    "\n",
    "### Running BLAST over the Internet\n",
    "\n",
    "This tool allows you to call the online version of BLAST and doesn't need you to download the databases to local. It can save you time and resources if the number of quries you are going to run isn't very large. \n",
    "\n",
    "We use the function `qblast()` in the `Bio.Blast.NCBIWWW` module to call the online version of BLAST. This function has three non-optional arguments:\n",
    "\n",
    "* The first argument is the __blast program__ to use for the search, as a lower case string. Currently `qblast` only works with blastn, blastp, blastx, tblast, and tblastx. \n",
    "* The second argument specifies the __databases__ to search against. The options for this are available on the [NCBI Guide to BLAST](https://ftp.ncbi.nlm.nih.gov/pub/factsheets/HowTo_BLASTGuide.pdf).\n",
    "* The third argument is a string containing your __query sequence__. This can either be the sequence itself, the sequence in fasta format, or an identifier like a GI number. \n",
    "\n",
    "__Usage Guidelines:__\n",
    "\n",
    "The NCBI BLAST servers are a shared resource. It gives priority to interactive users. In order to ensure availability of the service to the entire community, NCBI may limit searches for some high volume users and will move searches of users who submit more than 100 searches in a 24 hour period to a slower queue, or, in extreme cases, will block the requests. To avoid problems, API users should comply with the following guidelines:\n",
    "\n",
    "* Do not contact the server more often than once every 10 seconds.\n",
    "* Do not poll for any single Request ID more often than once a minute. \n",
    "* Use the URL parameter email and tool, so that the NCBI can contact you if there is a problem.\n",
    "* Run scripts weekends or between 9pm and 5am Eastern Time on weekdays if more than 50 searches will be submitted. \n",
    "\n",
    "To fulfill the third point, one can set the `NCBIWWW.email` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "587f8f12",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "\n",
    "NCBIWWW.email = \"u1133824@anu.edu.au\" "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c7938c4",
   "metadata": {},
   "source": [
    "The `qblast` function also takes a number of other option arguments, which are basically analogous to the different parameters you can set on the BLAST web page. We'll highlight a few of them here:\n",
    "\n",
    "* The argument `url_base` sets the base URL for running BLAST over the internet. By default it connects to the NCBI, but one can use this to connect to an instance of NCBI BLAST running in the cloud. \n",
    "* The `qblast` function can return the BLAST results in various formats, which you can choose with the optional `format_type` keyword: \"HTML\", \"Text\", \"ASN.1\", or \"XML\". The default is \"XML\", as that is the format expected by the parser. \n",
    "* The argument `expect` sets the expectation or e-value threshold.\n",
    "\n",
    "For more about the optional BLAST arguments, you can read the NCBI BLAST's own documentation, or the help documentation built into Biopython:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "31740c81",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function qblast in module Bio.Blast.NCBIWWW:\n",
      "\n",
      "qblast(program, database, sequence, url_base='https://blast.ncbi.nlm.nih.gov/Blast.cgi', auto_format=None, composition_based_statistics=None, db_genetic_code=None, endpoints=None, entrez_query='(none)', expect=10.0, filter=None, gapcosts=None, genetic_code=None, hitlist_size=50, i_thresh=None, layout=None, lcase_mask=None, matrix_name=None, nucl_penalty=None, nucl_reward=None, other_advanced=None, perc_ident=None, phi_pattern=None, query_file=None, query_believe_defline=None, query_from=None, query_to=None, searchsp_eff=None, service=None, threshold=None, ungapped_alignment=None, word_size=None, short_query=None, alignments=500, alignment_view=None, descriptions=500, entrez_links_new_window=None, expect_low=None, expect_high=None, format_entrez_query=None, format_object=None, format_type='XML', ncbi_gi=None, results_file=None, show_overview=None, megablast=None, template_type=None, template_length=None)\n",
      "    BLAST search using NCBI's QBLAST server or a cloud service provider.\n",
      "    \n",
      "    Supports all parameters of the old qblast API for Put and Get.\n",
      "    \n",
      "    Please note that NCBI uses the new Common URL API for BLAST searches\n",
      "    on the internet (http://ncbi.github.io/blast-cloud/dev/api.html). Thus,\n",
      "    some of the parameters used by this function are not (or are no longer)\n",
      "    officially supported by NCBI. Although they are still functioning, this\n",
      "    may change in the future.\n",
      "    \n",
      "    The Common URL API (http://ncbi.github.io/blast-cloud/dev/api.html) allows\n",
      "    doing BLAST searches on cloud servers. To use this feature, please set\n",
      "    ``url_base='http://host.my.cloud.service.provider.com/cgi-bin/blast.cgi'``\n",
      "    and ``format_object='Alignment'``. For more details, please see\n",
      "    https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=CloudBlast\n",
      "    \n",
      "    Some useful parameters:\n",
      "    \n",
      "     - program        blastn, blastp, blastx, tblastn, or tblastx (lower case)\n",
      "     - database       Which database to search against (e.g. \"nr\").\n",
      "     - sequence       The sequence to search.\n",
      "     - ncbi_gi        TRUE/FALSE whether to give 'gi' identifier.\n",
      "     - descriptions   Number of descriptions to show.  Def 500.\n",
      "     - alignments     Number of alignments to show.  Def 500.\n",
      "     - expect         An expect value cutoff.  Def 10.0.\n",
      "     - matrix_name    Specify an alt. matrix (PAM30, PAM70, BLOSUM80, BLOSUM45).\n",
      "     - filter         \"none\" turns off filtering.  Default no filtering\n",
      "     - format_type    \"HTML\", \"Text\", \"ASN.1\", or \"XML\".  Def. \"XML\".\n",
      "     - entrez_query   Entrez query to limit Blast search\n",
      "     - hitlist_size   Number of hits to return. Default 50\n",
      "     - megablast      TRUE/FALSE whether to use MEga BLAST algorithm (blastn only)\n",
      "     - short_query    TRUE/FALSE whether to adjust the search parameters for a\n",
      "                      short query sequence. Note that this will override\n",
      "                      manually set parameters like word size and e value. Turns\n",
      "                      off when sequence length is > 30 residues. Default: None.\n",
      "     - service        plain, psi, phi, rpsblast, megablast (lower case)\n",
      "    \n",
      "    This function does no checking of the validity of the parameters\n",
      "    and passes the values to the server as is.  More help is available at:\n",
      "    https://ncbi.github.io/blast-cloud/dev/api.html\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(NCBIWWW.qblast)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bedb141",
   "metadata": {},
   "source": [
    "Note that the default settings on the NCBI BLAST website are not quite the same as the defaults on QBLAST. If you get different results, you'll need to check the parameters (e.g., the expectation value threshold and the gap values). \n",
    "\n",
    "__Code Example 01:__ using GI number of your query sequence to search against the nucleotide database (nt) using BLASTN."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "43a5b885",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", \"8332116\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f61caaf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(result_handle.getvalue())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "560f7668",
   "metadata": {},
   "source": [
    "The result we get from `qblast` is a StringIO object. A StringIO object is a text stream using an in-memory text buffer. The `getvalue()` method of `StringIO` will return a string containing the entire contents of the buffer. \n",
    "\n",
    "__Code Example 02:__ if we have our query sequence already in a FASTA formatted file, we just need to open the file and read in this record as a string, and use that as the query argument:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6b8be4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_string = open(\"data/one_cds.fa\").read()\n",
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", fasta_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7f10415",
   "metadata": {},
   "source": [
    "__Code Example 03:__ we could also have read in the FASTA file as a `SeqRecord` object and then supplied just the sequence itself:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c833569",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b1038336",
   "metadata": {},
   "source": [
    "### Saving BLAST Output  \n",
    "\n",
    "Whatever arguments you give the `qblast()` function, you should get back your results in a handle object (by default in XML format). To save a handle object, we can use the following code to save it to file \"my_blast.xml\". "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86a9d56d",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/my_blast.xml\", \"w\") as out_handle:\n",
    "    out_handle.write(result_handle.read())\n",
    "\n",
    "result_handle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d28e26f9",
   "metadata": {},
   "source": [
    "`result_handle.close()` closes the handle object and the handle can no longer be read. We can try to get the value of the result_handle again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0c43310",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_handle.getvalue()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a44e7edd",
   "metadata": {},
   "source": [
    "### Parsing BLAST Output \n",
    "\n",
    "Biopython recommend to parse the output in XML format because it is more stable than the plain text and HTML output and much easier to parse automatically. You can get BLAST output in XML format in various ways. For the parser, it doesn't matter how the output was generated, as long as it is in the XML format. \n",
    "\n",
    "To input the XML result from file into Python environment, we use the `open()` function. It will read the file into a handle. In Python, a handle is a nice way of describing input to any info source so that the info can be retrieved using `read()` and `readline()` functions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "390a1bd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_handle = open(\"data/my_blast.xml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e16bdd9",
   "metadata": {},
   "source": [
    "Now that we have input the XML result into a handle again, we are ready to parse the output. The code to parse BLAST result of a single query is different from the result of multiple queries. \n",
    "\n",
    "__Parsing BLAST result for a single query:__\n",
    "\n",
    "The `NCBIXML.read()` function will read from the handle and store the parsed information in a Blast record object. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78f00dca",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "blast_record = NCBIXML.read(result_handle)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "518cfc07",
   "metadata": {},
   "source": [
    "__Parsing BLAST result for multiple queries:__\n",
    "\n",
    "The `NCBIXML.parse()` function can return the BLAST result as an iterator, you can use for loop to access each of the BLAST result or store them in a list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa06cfbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "result_handle = open(\"data/two_cds_result.xml\")\n",
    "blast_records = NCBIXML.parse(result_handle)\n",
    "blast_records_list = list(blast_records)\n",
    "\n",
    "len(blast_records_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85035728",
   "metadata": {},
   "source": [
    "__Exercise: please check what attributes and methods does this BLAST record class have?__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "979b855f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bd8c61e5",
   "metadata": {},
   "source": [
    "### The BLAST Record Class \n",
    "\n",
    "A BLAST Record contains everything you might ever want to extract from the BLAST output. Here, we will show an example of how to get some info out of the BLAST report.\n",
    "\n",
    "To continue with our example result from GI number `8332116`, let's print out some summary info about all hits in our blast report greater than a particular threshold. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e0dcac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "E_VALUE_THRESH = 0.04\n",
    "\n",
    "for alignment in blast_record.alignments:\n",
    "    for hsp in alignment.hsps:\n",
    "        if hsp.expect < E_VALUE_THRESH:\n",
    "            print(\"****Alignment****\")\n",
    "            print(\"sequence:\", alignment.title)\n",
    "            print(\"length:\", alignment.length)\n",
    "            print(\"e value:\", hsp.expect)\n",
    "            print(hsp.query[0:75] + \"...\")\n",
    "            print(hsp.match[0:75] + \"...\")\n",
    "            print(hsp.sbjct[0:75] + \"...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66f73e2b",
   "metadata": {},
   "source": [
    "Basically, you can do anything you want to with the info in the BLAST report once you have parsed it. You just have to spend more time on understanding the BLAST record object and how to extract the information you need. These objects are defined in `Bio.Blast.Record` and are quite complete. You can find the documentation [here](https://biopython.org/docs/1.75/api/Bio.Blast.Record.html). \n",
    "\n",
    "## More Tools in Biopython\n",
    "\n",
    "Biopython is a very powerful package and it covers a wide range of biological questions. There are many more things we can do using Biopython, and I will introduce a few here. If you would like to know more about Biopython, you can read their [Tutorial and Cookbook](http://biopython.org/DIST/docs/tutorial/Tutorial.html) combined with detailed explanation of Biopython Modules here [API Documentation](https://biopython.org/docs/1.75/api/index.html). \n",
    "\n",
    "### Connecting with biological databases \n",
    "\n",
    "One of the very common things that we need to do in bioinformatics is extract information from biological databases. It can be quite tedious to access these databases manually, especially if you have a lot of repetitive work to do. Biopython attempts to save you time and energy by making some online databases available from Python scripts. Currently, Biopython has code to extract information from the following databases:\n",
    "\n",
    "* Entrez (and PubMed) from the NCBI\n",
    "* ExPASy \n",
    "* SCOP \n",
    "\n",
    "### Going 3D: The PDB Module\n",
    "\n",
    "`Bio.PDB` is a Biopython module that focuses on working with crystal structures of biological macromolecules. \n",
    "\n",
    "### Bio.PopGen: Population Genetics\n",
    "\n",
    "`Bio.PopGen` is a Biopython module supporting population genetics. The objective for the module is to support widely used data formats, applications, and databases. `Bio.PopGen` provides a parser and generator of GenePop file format. Utilities to manipulate the content of a record are also provided. \n",
    "\n",
    "### Phylogenetics with Bio.Phylo\n",
    "\n",
    "The `Bio.Phylo` module aims to provide a common way to work with phylogenetic trees independently of the source data format, as well as a consistent API for I/O operations. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccbb2078",
   "metadata": {},
   "source": [
    "# References\n",
    "\n",
    "* Biopython - [Biopython Tutorial and Cookbook](http://biopython.org/DIST/docs/tutorial/Tutorial.html) \n",
    "* Wikipedia - [BLAST (biotechnology)](https://en.wikipedia.org/wiki/BLAST_(biotechnology))\n",
    "* NCBI - [How To BLAST Guide](https://ftp.ncbi.nlm.nih.gov/pub/factsheets/HowTo_BLASTGuide.pdf) \n",
    "* NCBI - [BLAST Help Developer Information](https://blast.ncbi.nlm.nih.gov/doc/blast-help/developerinfo.html#developerinfo)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
